File: internal/models/latency.go
package models

import (
	"slices"
)

func CalculateLatencyStats(latencies []int64) LatencyStats {
	if len(latencies) == 0 {
		return LatencyStats{}
	}

	sorted := make([]int64, len(latencies))
	copy(sorted, latencies)
	slices.Sort(sorted)

	var sum int64
	for _, lat := range sorted {
		sum += lat
	}

	return LatencyStats{
		P50: Percentile(sorted, 50),
		P90: Percentile(sorted, 90),
		P95: Percentile(sorted, 95),
		P99: Percentile(sorted, 99),
		Min: sorted[0],
		Max: sorted[len(sorted)-1],
		Avg: sum / int64(len(sorted)),
	}
}

func Percentile(latencies []int64, p int) int64 {
	if len(latencies) == 0 {
		return 0
	}

	idx := (len(latencies) * p / 100)
	if idx >= len(latencies) {
		idx = len(latencies) - 1
	}

	return latencies[idx]
}

File: crates/replayer-core/src/latency.rs
use crate::models::LatencyStats;

pub fn calculate_latency_stats(latencies: &[i64]) -> LatencyStats {
    if latencies.is_empty() {
        return LatencyStats::default();
    }

    let mut sorted: Vec<i64> = latencies.to_vec();
    sorted.sort_unstable();

    let sum: i64 = sorted.iter().sum();

    LatencyStats {
        p50: percentile(&sorted, 50),
        p90: percentile(&sorted, 90),
        p95: percentile(&sorted, 95),
        p99: percentile(&sorted, 99),
        min: sorted[0],
        max: sorted[sorted.len() - 1],
        avg: sum / sorted.len() as i64,
    }
}

pub fn percentile(latencies: &[i64], p: usize) -> i64 {
    if latencies.is_empty() {
        return 0;
    }

    let mut idx = latencies.len() * p / 100;
    if idx >= latencies.len() {
        idx = latencies.len() - 1;
    }

    latencies[idx]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_percentile_empty() {
        assert_eq!(percentile(&[], 50), 0);
    }

    #[test]
    fn test_percentile_single() {
        assert_eq!(percentile(&[42], 50), 42);
        assert_eq!(percentile(&[42], 99), 42);
    }

    #[test]
    fn test_calculate_latency_stats() {
        let latencies: Vec<i64> = (1..=100).collect();
        let stats = calculate_latency_stats(&latencies);
        assert_eq!(stats.min, 1);
        assert_eq!(stats.max, 100);
        assert_eq!(stats.avg, 50);
        assert_eq!(stats.p50, 51);
        assert_eq!(stats.p90, 91);
        assert_eq!(stats.p95, 96);
        assert_eq!(stats.p99, 100);
    }

    #[test]
    fn test_calculate_latency_stats_empty() {
        let stats = calculate_latency_stats(&[]);
        assert_eq!(stats.min, 0);
        assert_eq!(stats.max, 0);
        assert_eq!(stats.avg, 0);
    }
}
